TestModels/Union/src/SimpleResource.dfy

```diff
@@ -0,0 +1,135 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+include "../Model/SimpleUnionTypes.dfy"
+
+module SimpleResource {
+  import opened Types = SimpleUnionTypes
+  import opened Wrappers
+
+  /*
+    It is hard to prove things about mutable state.
+    Smithy-Dafny attempts to help with this
+    by adding the requires, ensures, and modifies clauses.
+    However making sure that this is generated correctly
+    can be complicated and tests are a very good idea.
+
+    Further it is possible to generate a Dafny specification
+    that can never be used.
+    So this file is to make sure that the code is generates
+    and verified in both the Dafny types file,
+    and in an implementation.
+    As well as verify at least one implementation is callable.
+  */
+
+  class SomeResource
+    extends Types.ISimpleResource
+    {
+
+    predicate ValidState()
+      ensures ValidState() ==> History in Modifies
+    {
+      && History in Modifies
+    }
+
+    predicate GetResourceDataEnsuresPublicly(input: GetResourceDataInput , output: Result<GetResourceDataOutput, Error>)
+    {
+      true
+    }
+
+    constructor() 
+      ensures fresh(Modifies)
+      ensures ValidState()
+    {
+      History := new Types.ISimpleResourceCallHistory();
+      Modifies := {History};
+    }
+
+    method GetResourceData' ( input: GetResourceDataInput )
+      returns (output: Result<GetResourceDataOutput, Error>)
+      requires
```

Comment:
Musing: boy is it unfortunate that you have you copy the spec here, especially when you don't have to for service operations (since module refinement doesn't make you copy the spec) :P

Reply:
I know! It has come up before that such things should just get inserted in the same way with traits...